Provide quick info at #regionend showing #region message

MaStr11 opened the pull request:
Fixes #56507

MaStr11 commented on src/Features/CSharp/Portable/QuickInfo/CSharpSyntacticQuickInfoProvider.cs:
```diff
@@ -28,18 +29,30 @@ public CSharpSyntacticQuickInfoProvider()
             {
-                return null;
+                case SyntaxKind.CloseBraceToken:
+                    break;
```
The  code following this switch statement is handling the Brace cases. I would refactor this into that into `BuildQuickInfoCloseBrace`. It might be even useful to place the functions `BuildQuickInfoEndRegion` and `BuildQuickInfoCloseBrace` into their own partial classes. What do you think?

MaStr11 commented on src/Features/CSharp/Portable/QuickInfo/CSharpSyntacticQuickInfoProvider.cs:
```diff
@@ -28,18 +29,30 @@ public CSharpSyntacticQuickInfoProvider()
+                case SyntaxKind.HashToken:
+                case SyntaxKind.EndOfDirectiveToken:
+                    return BuildQuickInfoEndRegion(token, cancellationToken);
+                case SyntaxKind.EndOfFileToken:
+                    var endRegionKeyword = token.LeadingTrivia.LastOrDefault().GetStructure()?.ChildTokens().FirstOrDefault(t => t.IsKind(SyntaxKind.EndRegionKeyword));
+                    return endRegionKeyword.HasValue
+                        ? BuildQuickInfoEndRegion(endRegionKeyword.Value, cancellationToken)
+                        : null;
```
I'm not sure if the EOF case is worth to handle. If so, we might try to find something better than `token.LeadingTrivia.LastOrDefault().GetStructure()?.ChildTokens().FirstOrDefault(t => t.IsKind(SyntaxKind.EndRegionKeyword))`

MaStr11 commented on src/Features/CSharp/Portable/QuickInfo/CSharpSyntacticQuickInfoProvider.cs:
```diff
@@ -79,6 +92,20 @@ public CSharpSyntacticQuickInfoProvider()
 
+        private static QuickInfoItem? BuildQuickInfoEndRegion(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            if (token.Parent is EndRegionDirectiveTriviaSyntax endRegionDirectiveTrivia)
+            {
+                var regionStart = endRegionDirectiveTrivia.GetMatchingDirective(cancellationToken);
```
This would also work for
```CS
#if wouldBeShownFor
#elif
#else
#endif
```
Should we consider the `#if` directives too?

MaStr11 replied:
Already done in 077db26

MaStr11 commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/SyntaxTreeExtensions.cs:
```diff
@@ -68,7 +68,7 @@ public static bool IsScript(this SyntaxTree syntaxTree)
             Contract.ThrowIfNull(syntaxTree);
 
-            if (position >= syntaxTree.Length)
+            if (position > syntaxTree.Length)
```
`position >= syntaxTree.Length` misses EOF.

MaStr11 commented on src/Features/CSharp/Portable/QuickInfo/CSharpSyntacticQuickInfoProvider.cs:
```diff
@@ -28,20 +30,38 @@ public CSharpSyntacticQuickInfoProvider()
+                case SyntaxKind.ElifKeyword:
+                case SyntaxKind.EndOfDirectiveToken:
+                    return BuildQuickInfoDirectives(token, cancellationToken);
+                case SyntaxKind.EndOfFileToken:
+                    var endRegionKeyword = token.LeadingTrivia.LastOrDefault().GetStructure()?.ChildTokens().FirstOrDefault(t => t.IsKind(SyntaxKind.EndRegionKeyword));
+                    return endRegionKeyword.HasValue
+                        ? BuildQuickInfoDirectives(endRegionKeyword.Value, cancellationToken)
+                        : null;
```
I think this should be dropped, because it just complicates things without much gain.

CyrusNajmabadi replied:
what cases is it needed for?

MaStr11 replied:
Just an `#endregion$$` with EOF following.

MaStr11 replied:
I could take a commit and remove the test cases that fail after the removal

MaStr11 replied:
I added commit 6b670383491f9e0fc4ef9788e1688e3463c5fd05
There I removed the EOF handling and the removed test cases failed after that changed. I can revert the change if you think that the cases are important to support.

CyrusNajmabadi commented on src/Features/CSharp/Portable/QuickInfo/CSharpSyntacticQuickInfoProvider.cs:
```diff
@@ -28,20 +30,38 @@ public CSharpSyntacticQuickInfoProvider()
+                case SyntaxKind.ElifKeyword:
+                case SyntaxKind.EndOfDirectiveToken:
+                    return BuildQuickInfoDirectives(token, cancellationToken);
+                case SyntaxKind.EndOfFileToken:
```
when is this case hit?

MaStr11 replied:
e.g. in some of the RegionEndShowsStartRegionMessageAtDifferentPositions cases:
```CS
#region Start
#endregion$$
```
But I would like to remove that case because I don't thinks it is worth the complications.

MaStr11 commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/SyntaxTreeExtensions.cs:
```diff
@@ -68,22 +68,27 @@ public static bool IsScript(this SyntaxTree syntaxTree)
             var token = root.FindToken(position, findInsideTrivia);
-
-            if ((token.Span.Contains(position) || token.Span.End == position) && predicate(token))
+            var span = findInsideTrivia
+                ? token.FullSpan
+                : token.Span;
+            if ((span.Contains(position) || span.End == position) && predicate(token))
```
The "FullSpan" is required for (at least) these cases:
```
$$ #endregion
# $$ endregion
#endregion En$$d
```

MaStr11 commented on src/EditorFeatures/CSharpTest/QuickInfo/SyntacticQuickInfoSourceTests.cs:
```diff
@@ -265,6 +265,254 @@ public async Task ScopeBrackets_10()
+#elif RELEASE
+#elif DEMO
+#el$$se
+#endif", "#if DEBUG#elif RELEASE#elif DEMO");
```
These are three different "relatedSpans". I don't know how QuickInfo handles such cases.

CyrusNajmabadi replied:
can you check and show a picture of what this looks like?  thanks! :)

MaStr11 replied:
I can't, unfortunately. I'm not able to F5 because of the VS 2022 preview requirement. Does it help to chow you the QuickInfoItem definition for this case?

CyrusNajmabadi replied:
NOte: 2022 can be sxs installed with 2019.  That said, i'll pull this down locally to try out.

MaStr11 replied:
I know. But I'm not allowed to do so. Thanks for taking the time to try this out!

CyrusNajmabadi reviewed (approved):
Were you planning on doing VB as well?

MaStr11 commented:
> Were you planning on doing VB as well?

This could be easily be done if GetMatchingConditionalDirectives() and GetMatchingDirective() are implemented for VB. I haven't checked that. So the answer would "Maybe in a follow-up PR".

CyrusNajmabadi commented:
They should be implemented already.  I believe they're there for commands like "goto matching brace"

CyrusNajmabadi commented:
I'm also ok with VB support coming later based on demand.

CyrusNajmabadi commented:
Yeah, i genuinely am not a fan of htis:

![image](https://user-images.githubusercontent.com/4564579/134544782-5cb65a8a-6ef8-44c4-bf37-7518c4a5951c.png)

I think we shoudl either limit ourselves to the first span.  Or, if we're including all of them, we include the end-of-line token at the end of the directive so that doesn't show up concatenated.

MaStr11 commented:
Agreed. This looks awful and isn't what I was expecting. I will look into this next week. So it might be good idea to remove your approval until this is fixed.

MaStr11 commented:
> Yeah, i genuinely am not a fan of htis:

![](https://user-images.githubusercontent.com/4564579/134544782-5cb65a8a-6ef8-44c4-bf37-7518c4a5951c.png)

>  I think we shoudl either limit ourselves to the first span. Or, if we're including all of them, we include the end-of-line token at the end of the directive so that doesn't show up concatenated.

I tried to include the end-of-line token, but it is ignored/dropped by the `IntellisenseQuickInfoBuilder.BuildItemAsync` logic. I further tried to build `TaggedText` by hand and pass it via the `QuickInfoItem.Create(sections)` argument, but that was a death end, because `TaggedText` does not support preprocessor directives (There is a `ClassificationTypeNames.PreprocessorKeyword` but no corresponding `TextTags`).

I don't see any way to include a new line between the directives, without changing some QuickInfo building blocks (The most reasonable change seems to be to include a newLine, if "realtedSpans" contains non-successive spans).

The easiest way forward is, to follow your first suggestion:

>  I think we shoudl either limit ourselves to the first span.

I did this in cf37ecf (but I included the last preceding `#if` or `#elif` instead of the first one).

MaStr11 commented:
This test is failing now:
https://github.com/dotnet/roslyn/blob/40cfa4b8ff85eb0c37622fca8bdafb97521168f8/src/Features/LanguageServer/ProtocolUnitTests/Hover/HoverTests.cs#L117-L137

The regression was introduced in e78e895ee4d44a3d2c9859fd315b23d5644a69d0. "Hover" covers the "FullSpan" now, and therefore QuickInfo is shown for the closing brace (showing the opening brace including the method signature).

@CyrusNajmabadi I can't tell if the test assert need to be changed or if e78e895ee4d44a3d2c9859fd315b23d5644a69d0 needs to be reverted.

MaStr11 commented:
@CyrusNajmabadi Notable changes since your last review:
* Only the most recent `#if` or `#elif` is shown (see https://github.com/dotnet/roslyn/pull/56609/#issuecomment-927721391 for details)
* Some edge cases are no longer supported, see [RegionEndQuickInfoIsNotOfferedAtDifferentPositions](https://github.com/dotnet/roslyn/pull/56609/files#diff-765b6cffd9c3daf9cf85b958fe2d210c289c278b9ce20a415a962a91680843d7R293-R301) for examples of unsupported cases
* A lot more tokens are taken into account and `token.Parent.AncestorsAndSelf` is needed to support Quickinfo in `#elif con$$dition` see [ElifHasQuickinfoAtDifferentPositions](https://github.com/dotnet/roslyn/pull/56609/files#diff-765b6cffd9c3daf9cf85b958fe2d210c289c278b9ce20a415a962a91680843d7R525-R541) for details.
* Previous changes to token handling in different files were reverted (they introduced regressions elsewhere).

CyrusNajmabadi commented:
@MaStr11 is this ready for review?

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/IntellisenseQuickInfoBuilder.cs:
```diff
@@ -100,17 +100,31 @@ internal static class IntellisenseQuickInfoBuilder
+                    {
+                        // If there is more than one relatedSpan, the single spans need to be separated.
+                        // This marker is used to track, where the separator needs to be inserted.
+                        classifiedSpanList.Add(new ClassifiedSpan(span, RelatedSpanSeparatorClassification));
```
is this still used in this pr?  when do you have more htan one related span?

MaStr11 replied:
That's when you have `#if #elif #endi$$f` I reintroduced to show the whole conditionals today. This change needs to be verified by you again, because I can't F5 and screenshot it.

CyrusNajmabadi commented on src/Features/CSharp/Portable/QuickInfo/CSharpSyntacticQuickInfoProvider.cs:
```diff
@@ -28,20 +30,43 @@ public CSharpSyntacticQuickInfoProvider()
+                case SyntaxKind.EndOfDirectiveToken:
+                // Allowed tokens in #elif condition 
+                case SyntaxKind.IdentifierToken:
+                case SyntaxKind.AmpersandAmpersandToken:
```
i don't really get the point of supporting && and || and whatnot.  i would expect this to only really work on the `#keyword` portion, not the subexpression portino.

MaStr11 replied:
Agreed. I will remove it.

MaStr11 replied:
Done in a7c638f

CyrusNajmabadi commented on src/Features/CSharp/Portable/QuickInfo/CSharpSyntacticQuickInfoProvider.cs:
```diff
@@ -125,5 +150,34 @@ private static bool TryFindFurthestNearbyComment<T>(ref T triviaSearchList, out
+                    .TakeWhile(d => d.SpanStart < directiveTrivia.SpanStart)
+                    .Select(d => d.Span)
+                    .ToImmutableArray();
+                if (lastMatchBefore.Length > 0)
```
this is named weird.  it says 'lastMatchBefore'.  but it matches multiple things.

CyrusNajmabadi commented on src/Features/CSharp/Portable/QuickInfo/CSharpSyntacticQuickInfoProvider.cs:
```diff
@@ -125,5 +150,34 @@ private static bool TryFindFurthestNearbyComment<T>(ref T triviaSearchList, out
+
+        private static QuickInfoItem? BuildQuickInfoDirectives(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            var directiveTrivia = token.Parent?.AncestorsAndSelf(ascendOutOfTrivia: false).FirstOrDefault(n =>
+                n is EndRegionDirectiveTriviaSyntax or ElifDirectiveTriviaSyntax or ElseDirectiveTriviaSyntax or EndIfDirectiveTriviaSyntax) as DirectiveTriviaSyntax;
```
```suggestion
            var directiveTrivia = token.Parent?.AncestorsAndSelf(ascendOutOfTrivia: false).OfType<DirectiveTriviaSyntax>().FirstOrDefault();
```

CyrusNajmabadi commented on src/Features/CSharp/Portable/QuickInfo/CSharpSyntacticQuickInfoProvider.cs:
```diff
@@ -125,5 +150,34 @@ private static bool TryFindFurthestNearbyComment<T>(ref T triviaSearchList, out
+                var matchingDirectives = directiveTrivia.GetMatchingConditionalDirectives(cancellationToken);
+                var lastMatchBefore = matchingDirectives
+                    .TakeWhile(d => d.SpanStart < directiveTrivia.SpanStart)
+                    .Select(d => d.Span)
```
instead of having the quick info code have to add \r\n's as it's combining related spans, why not just include the \r\n portion in the span you're grabbing here.  tehn it should happen automatically, right?

MaStr11 replied:
I tried this, but the newLine token got dismissed in the QuickInfo builder logic. I even tried to use the `sections` parameter but `TaggedText` does not support directives. See also https://github.com/dotnet/roslyn/pull/56609#issuecomment-927721391

CyrusNajmabadi commented:
@MaStr11 if you don't want to make the whitespace changes, i'm happy to do it for this PR.  thanks!

MaStr11 commented on src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/IntellisenseQuickInfoBuilder.cs:
```diff
@@ -100,17 +100,31 @@ internal static class IntellisenseQuickInfoBuilder
+                for (var i = 0; i < quickInfoItem.RelatedSpans.Length; i++)
                 {
+                    var span = quickInfoItem.RelatedSpans[i];
                     var classifiedSpans = await ClassifierHelper.GetClassifiedSpansAsync(document, span, cancellationToken).ConfigureAwait(false);
```
There are several building blocks that don't work well with the newLine. GetClassifiedSpansAsync is one of them, but the following GetSpansWithAlignedIndentation might interfere here too.

MaStr11 replied:
> if the related span included the \r\n part of the file, then it seems like GetClassifiedSpansAsync should work here

I made this work in a498d5a.

MaStr11 commented:
@CyrusNajmabadi I fixed the wrong naming and removed the #elif condition support. You can take over and fix the "newLine" problem if you like. (Hint: we can also go back to commit 7300c96, there I only included the last #if or #elif before the cursor, so the newLine problem is irrelevant).

MaStr11 commented:
@CyrusNajmabadi I fixed the newLine problem in a498d5a. Please check if this is the desired solution. If it's not, you are welcome to take over. Please keep in mind, that you still need to pull and F5 the PR to check, that the newLine is indeed respected in the QuickInfo.

If you want, I could add some more tests for the newLine problem, to make sure it behaves as expected e.g. for intended directives.

MaStr11 commented:
> if the related span included the \r\n part of the file, then it seems like GetClassifiedSpansAsync should work here

I tried three different approaches now

1. Add new newlLine between relatedSpans (This was my original solution and I did this slightly different in the latest version again).
2. Change the classification worker to include trailing trivia (see commit a498d5afe263ce624bb911a37fb4ac5e8d3e51b1, which caused lots of test failures in the classification test suite. see https://dev.azure.com/dnceng/public/_build/results?buildId=1395137&view=ms.vss-test-web.build-test-results-tab)
3. Only show a single relatedSpan for conditionals (see cf37ecf9d61bf21b1a823426308d165c365fcc1e)

I think 1. is the most reasonable approach. If there is more than one relatedSpan, the spans need to be separated somehow. The separator should be passed to `QuickInfoItem.Create` to give the caller more control, but I hard-coded a line break at the moment.

CyrusNajmabadi commented:
Thanks @MaStr11 !
